ast/src/parsed/asm.rs

```diff
@@ -44,42 +44,51 @@ impl Params {
     }
 }
 
-#[derive(Debug, PartialEq, Eq, Clone)]
-/// the function id necessary to call this function from the outside
-pub struct FunctionId<T> {
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+/// the operation id necessary to call this function from the outside
+pub struct OperationId<T> {
     pub id: T,
 }
 
-#[derive(Debug, PartialEq, Eq, Clone)]
+#[derive(Debug, PartialEq, Eq, Clone, PartialOrd, Ord)]
 pub struct Instruction<T> {
     pub params: Params,
     pub body: InstructionBody<T>,
 }
 
-#[derive(Debug, PartialEq, Eq, Clone)]
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
 pub enum MachineStatement<T> {
     Degree(usize, AbstractNumberType),
     Submachine(usize, String, String),
     RegisterDeclaration(usize, String, Option<RegisterFlag>),
     InstructionDeclaration(usize, String, Instruction<T>),
+    LinkDeclaration(LinkDeclaration<T>),
     InlinePil(usize, Vec<PilStatement<T>>),
-    FunctionDeclaration(
-        usize,
-        String,
-        // function id, explicit for static machines
-        Option<FunctionId<T>>,
-        Params,
-        Vec<FunctionStatement<T>>,
-    ),
+    FunctionDeclaration(usize, String, Params, Vec<FunctionStatement<T>>),
+    OperationDeclaration(usize, String, OperationId<T>, Params),
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct LinkDeclaration<T> {
+    pub start: usize,
+    pub flag: Expression<T>,
+    pub params: Params,
+    pub to: CallableRef,
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct CallableRef {
+    pub instance: String,
+    pub callable: String,
```

Comment:
If this is a string, shouldn't the type checker check that the called operations has the correct arity?

Reply:
Yes it should. There is a lot more that can be checked in the type checker in general. I think it can be a different PR.

Reply:
ok, can you add a TODO in the type checker and open an issue

Reply:
added TODO to the checked AST (this is the parsed one where this doesnt have to be enforced)